probe/proc/cache.go

```
@@ -0,0 +1,123 @@
+	fce.Lock()
+	defer fce.Unlock()
+
+	if fce.Pinned() { // use the saved contents
+		buf.Write(fce.contents.Bytes())
+		return nil
+	}
+	return fce.file.ReadInto(buf)
+}
+
+// A cache for files handles
+type filesCache struct {
+	*lru.Cache
+	proc Dir
+	quit chan (struct{})
+}
+
+func newFilesCache(proc Dir) *filesCache {
+	evictionFunc := func(key, value interface{}) {
+		entry := value.(filesCacheEntry)
+		if entry.file != nil {
+			entry.file.Close()
+		}
+	}
+
+	cache, _ := lru.NewWithEvict(filesCacheLen, evictionFunc)
+	fc := filesCache{
+		Cache: cache,
+		proc:  proc,
+		quit:  make(chan struct{}),
+	}
+	go fc.loop()
+
+	return &fc
+}
+
+// Read a "/proc" file, identified as a sub-path (eg "1134/comm"), into a buffer
+// When the "pin" argument is true, we will cache the file _content_.
+func (fc *filesCache) ReadInto(filename string, buf *bytes.Buffer, pin bool) error {
+	e, ok := fc.Cache.Get(filename)
+	if !ok {
+		f, err := fc.proc.Open(filename)
+		if err != nil {
+			return err
+		}
+		entry := filesCacheEntry{
+			creation: time.Now(),
+		}
+		b := bytes.Buffer{}
+		if err := f.ReadInto(&b); err != nil {
+			return err
+		}
+		if pin {
+			entry.contents = &b // save the file contents
+			f.Close()           // and close the file
+		} else {
+			entry.file = f
+		}
+		buf.Write(b.Bytes())
+		fc.Cache.Add(filename, entry)
+		return nil
+	}
+
+	entry := e.(filesCacheEntry)
+	return entry.ReadInto(buf)
+}
+
+// Close closes all the handles in the cache
+func (fc *filesCache) Close() error {
+	fc.Cache.Purge()
+	fc.quit <- struct{}{}
+	return nil
+}
+
+// Go through all the entries in the cache and force the removal of
+// the ones that were created too long ago
+func (fc *filesCache) loop() {
+	purge := time.Tick(filesCacheExpiration)
+
+	for {
+		select {
+		case <-purge:
+			now := time.Now()
+			timeLimit := now.Add(-filesCacheExpiration)
+			for _, key := range fc.Cache.Keys() {
+				e, _ := fc.Cache.Peek(key)
+				entry := e.(filesCacheEntry)
```

Don't ignore the error...